### ISSUE TITLE
handleCloudKitErrorAs and .limitExceeded

### ISSUE THREAD

**Issue by jfgates** [evermeer/EVCloudKitDao#102]
Hi,

I am rewriting my code to handle .limitExceeded.  Apple states that when an app receives .limitExceeded it should split the operation in half and try again.   I'd like a way to detect this so that I can change the quantity of records used in half.  

Currently, handleCloudKitErrorAs returns .fail for .limitExceeded.  I'd like to suggest a change to either use a new enum (eg, .limitExceeded) or that .fail would return the errorCode.code to caller with .fail.  The later would cause every app that uses .fail to be modified to handle the error code.

In the former case (new enum for limit exceeded)  I think that .quotaExceeded should be split from .limitExceeded, so that .quotaExceeded would remain as a .fail.

Example of what I'm thinking:

    open static func handleCloudKitErrorAs(_ error: Error?, retryAttempt: Double = 1) -> HandleCloudKitErrorAs {
        // There is no error
        if error == nil {
            return .success
        }

        // Or if there is a retry delay specified in the error, then use that.
        if let userInfo = error?._userInfo as? NSDictionary {
            if let retry = userInfo[CKErrorRetryAfterKey] as? NSNumber {
                let seconds = Double(retry)
                NSLog("Debug: Should retry in \(seconds) seconds. \(error?.localizedDescription ?? "")")
                return .retry(afterSeconds: seconds)
            }
        }

        let errorCode: CKError = CKError(_nsError: error! as NSError)
        switch errorCode.code {
        case .notAuthenticated, .networkUnavailable, .networkFailure, .serviceUnavailable, .requestRateLimited, .zoneBusy, .resultsTruncated:
            // Probably handled by the userInfo[CKErrorRetryAfterKey] but if not, then:
            // Use an exponential retry delay which maxes out at half an hour.
            var seconds = Double(pow(2, Double(retryAttempt)))
            if seconds > 1800 {
                seconds = 1800
            }
            NSLog("Debug: Should retry in \(seconds) seconds. \(error?.localizedDescription ?? "")")
            return .retry(afterSeconds: seconds)
        case .unknownItem, .invalidArguments, .incompatibleVersion, .badContainer, .missingEntitlement, .permissionFailure, .badDatabase, .assetFileNotFound, .operationCancelled, .assetFileModified, .batchRequestFailed, .zoneNotFound, .userDeletedZone, .internalError, .serverRejectedRequest, .constraintViolation:
            NSLog("Error: \(error?.localizedDescription ?? "")")
            return .fail
        case .limitExceeded:
            NSLog("Warning: \(error?.localizedDescription ?? "")")
            return .limitExceeded // new value added to EVCloudKit HandleCloudKitErrorAs enum
        case .quotaExceeded:
            NSLog("Warning: \(error?.localizedDescription ?? "")")
            return .fail
        case .changeTokenExpired,  .serverRecordChanged:
            NSLog("Info: \(error?.localizedDescription ?? "")")
            return .recoverableError
        default:
            NSLog("Error: \(error?.localizedDescription ?? "")") //New error introduced in iOS...?
            return .fail
        }
    }

and code that uses the above, 

    static func processError ( _ retryCount : Double, _ limit : Int, _ items: [ASDataObject], _ error : Error,  completionClosure: @escaping ( _ success: Bool, _ error: Error? ) -> (), updateClosure:  @escaping ( _ retryCount : Double, _ limit : Int ) -> ())
    {
        switch EVCloudKitDao.handleCloudKitErrorAs ( error, retryAttempt: retryCount )
        {
        case .retry ( let timeToWait ):
            EVLog ( "ERROR in ASDataObject.processError: Can retry after \( timeToWait )" )
            Async.background ( after: timeToWait )
            {
                updateClosure ( retryCount, limit )
            }
        case .limitExceeded:
            updateClosure ( retryCount, limit / 2 )
        case .fail:
            completionClosure ( false, error )
        default: // For here there is no need to handle the .Success, and .RecoverableError
            break
        }
    }

<Edit>  
     This note, above, is what one gets when writing it in the middle of the night....  Leaving the above as is, I have an alternate .fail case that doesn't require changes to your library.

    static func processError ( _ retryCount : Double, _ limit : Int, _ items: [ASDataObject], _ error : Error,  completionClosure: @escaping ( _ success: Bool, _ error: Error? ) -> (), updateClosure:  @escaping ( _ retryCount : Double, _ limit : Int ) -> ())
    {
        switch EVCloudKitDao.handleCloudKitErrorAs ( error, retryAttempt: retryCount )
        {
        case .retry ( let timeToWait ):
            EVLog ( "ERROR in ASDataObject.processError: Can retry after \( timeToWait )" )
            Async.background ( after: timeToWait )
            {
                updateClosure ( retryCount, limit )
            }
        case .fail:
            let errorCode: CKError = CKError(_nsError: error as NSError)
            let code = errorCode.code
            if code == .limitExceeded
            {
                Async.background ( )
                {
                    updateClosure ( retryCount, limit / 2 )
                }
                return
            }
            completionClosure ( false, error )
        default: // For here there is no need to handle the .Success, and .RecoverableError
            break
        }
    }

The only advantage of making a change is it requires less work on those who use handleCloudKitErrorAs().
</Edit>

My example code is work in progress..I'm still unit testing this idea.

What do you think?  Let me know if you need anything more.

Joseph
fgates@me.com

**Comment by evermeer** [comment]
I think it all comes down to what would be the easiest solution for the most common situations. What would you think about this:

I can just add an extra helper extension for easy acces to the error code like this:

```
public extension Error {
    public var code: CKError.Code {
        get {
            return CKError(_nsError: self as NSError).code
        }
    }
}
```

You can then just access the error like this:
```
                case .fail:
                    if error.code == .limitExceeded {
                        
                    }
```
The you do have the easy access. Nobody needs to change anything. There would also be support for potential other situations where you want to do something specific with the error.

**Comment by evermeer** [comment]
Since it's just a small change that adds a convenient property I just added it and published it as version 3.5.2.

**Comment by jfgates** [comment]
Great solution.   And much less invasive than my first thoughts.  Thanks again!
